### PR TITLE
Add GitHub actions for PHPUnit tests.

### PR DESCRIPTION
**Pull request by tofumatt**
## Summary

Addresses issue #2368.

## Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: .github/workflows/php-tests.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+on:
+  pull_request:
+    branches:
+      - master
+      - develop
+    types:
+      - opened
+      - reopened
+      - synchronize
+      - ready_for_review
```
Comment (felixarntz) [inline .github/workflows/php-tests.yml:12]: @tofumatt Let's also run this on push, see other comments on #2537:
```suggestion
on:
  push:
    branches:
      - develop
      - master
  pull_request:
    types:
      - opened
      - reopened
      - synchronize
      - ready_for_review
```

File: .github/workflows/php-tests.yml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  pull_request:
+    types:
+      - opened
+      - reopened
+      - synchronize
+      - ready_for_review
+
+jobs:
+  php-tests:
+    name: Run PHP Tests
```
Comment (aaemnnosttv) [inline .github/workflows/php-tests.yml]: We can use values from the matrix to make the job name dynamic.

See here how I have this [on another project](https://github.com/aaemnnosttv/version-assets/blob/9480d7e7666341840e04d831a38223ddf71dc19b/.github/workflows/tests.yml).

File: .github/workflows/php-tests.yml [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        php: ['5.6', '7.4']
```
Comment (aaemnnosttv) [inline .github/workflows/php-tests.yml]: 5.6 is a special case, so this should go in the matrices `include` key since we don't want to generate a permutation of each WP version for it.

This would only be `[ '7.4' ]`.

File: .github/workflows/php-tests.yml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        wordpress: ['4.7', 'latest-stable', 'latest-stable (multisite)', 'nightly']
```
Comment (aaemnnosttv) [inline .github/workflows/php-tests.yml]: We should be able to use versions here that are passed to the test install script.

This should only be `latest` and `nightly`. 4.7 and latest with multisite should be added via `include`.

`multisite` itself would be part of the matrix and be set like `env: WP_MULTISITE: {{ matrix.multisite }}` (compressed for brevity). See a similar example on the other project.

File: .github/workflows/php-tests.yml [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    services:
+      mysql:
+        image: mysql:5.7
+        env:
+          MYSQL_DATABASE: wordpress
```
Comment (aaemnnosttv) [inline .github/workflows/php-tests.yml:23]: This part should get the added health check command and may need the port exposed as well otherwise the test may be using the mysql server on the runner.
See https://github.com/aaemnnosttv/version-assets/blob/9480d7e7666341840e04d831a38223ddf71dc19b/.github/workflows/tests.yml#L32

File: .github/workflows/php-tests.yml [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    steps:
+      - uses: actions/checkout@v2
+      - uses: shivammathur/setup-php@v2
+        with:
+          extensions: mbstring, intl, redis, pdo_mysql
```
Comment (aaemnnosttv) [inline .github/workflows/php-tests.yml:28]: I don't _think_ we need `intl` but we for sure don't need `redis`, and `pdo_mysql` should probably be `mysqli` instead.

File: .github/workflows/php-tests.yml [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+          php-version: ${{ matrix.php }}
+      - name: Start MySQL
+        run: sudo service mysql start
```
Comment (aaemnnosttv) [inline .github/workflows/php-tests.yml:31]: This is the mysql instance on the runner (which we probably shouldn't use because we have no control over its version), not the one in the container added above 🙂

File: .github/workflows/php-tests.yml [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+      - name: Enable multisite
+        if: ${{ matrix.wordpress == 'latest-stable-multisite' }}
+        run: |
+          echo "WP_MULTISITE=1" >> $GITHUB_ENV
```
Comment (aaemnnosttv) [inline .github/workflows/php-tests.yml]: As mentioned before, this can be set unconditionally in the job's `env` (or in the test step) based on the value in the matrix. By default we can set it to `0` if needed, but probably unset/blank is fine too since this needs to be truthy to enable multisite.

File: .github/workflows/php-tests.yml [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+      - name: Extract version from stable WordPress
+        if: ${{ matrix.wordpress == 'latest-stable' || matrix.wordpress == 'latest-stable-multisite' }}
+        run: |
+          curl -s http://api.wordpress.org/core/version-check/1.7/ > /tmp/wp-latest.json
+          echo "WP_VERSION=$(grep -o '"version":"[^"]*' /tmp/wp-latest.json | sed 's/"version":"//')" >> $GITHUB_ENV
```
Comment (aaemnnosttv) [inline .github/workflows/php-tests.yml]: We shouldn't need to do this since the test setup/install script will install the latest version if we use the version `latest`.

File: .github/workflows/php-tests.yml [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        env:
+          WP_VERSION: ${{ matrix.wordpress }}
+      - name: Set WordPress version
+        if: ${{ !(matrix.wordpress == 'latest-stable' || matrix.wordpress == 'latest-stable (multisite)') }}
+        run: echo "WP_VERSION=$WP_VERSION" >> $GITHUB_ENV
+        env:
+          WP_VERSION: ${{ matrix.wordpress }}
```
Comment (aaemnnosttv) [inline .github/workflows/php-tests.yml]: I don't think we need it via env either - it's only needed for passing to the `tests/bin/install-wp-tests.sh` script.

File: .github/workflows/php-tests.yml [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+          echo "::set-output name=dir::$(composer config cache-files-dir)"
+      - uses: actions/cache@v2
+        with:
+          path: ${{ steps.composer-cache.outputs.dir }}
+          key: ${{ runner.os }}-composer-${{ hashFiles('**/composer.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-composer-
+      - name: Composer Install
+        if: ${{ (matrix.wordpress == '4.7' && matrix.php == '5.6') || (matrix.wordpress == 'latest-stable' && matrix.php == '7.4') || (matrix.wordpress == 'latest-stable (multisite)' && matrix.php == '7.4') || (matrix.wordpress == 'nightly' && matrix.php == '7.4') }}
+        run: docker run --rm -v "$PWD:/app" -v "${{ steps.composer-cache.outputs.dir }}:/tmp/cache" composer install --no-interaction --no-progress
```
Comment (aaemnnosttv) [inline .github/workflows/php-tests.yml]: Not sure why this is needed – we don't conditionally run composer any differently between the current PHP test jobs on Travis.

File: .github/workflows/php-tests.yml [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+      - name: Set up PHP test data
+        if: ${{ (matrix.wordpress == '4.7' && matrix.php == '5.6') || (matrix.wordpress == 'latest-stable' && matrix.php == '7.4') || (matrix.wordpress == 'latest-stable (multisite)' && matrix.php == '7.4') || (matrix.wordpress == 'nightly' && matrix.php == '7.4') }}
+        run: tests/bin/install-wp-tests.sh wordpress_test root root 127.0.0.1 $WP_VERSION
```
Comment (aaemnnosttv) [inline .github/workflows/php-tests.yml:153]: We should be able to do this instead:

```suggestion
        run: tests/bin/install-wp-tests.sh wordpress_test root root 127.0.0.1 {{ matrix.wordpress }}
```

File: .github/workflows/php-tests.yml [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+      - name: Run Unit Tests
+        if: ${{ (matrix.wordpress == '4.7' && matrix.php == '5.6') || (matrix.wordpress == 'latest-stable' && matrix.php == '7.4') || (matrix.wordpress == 'latest-stable (multisite)' && matrix.php == '7.4') || (matrix.wordpress == 'nightly' && matrix.php == '7.4') }}
```
Comment (aaemnnosttv) [inline .github/workflows/php-tests.yml]: This line should always run. Why would the main step run conditionally?

```suggestion
```

File: tests/bin/install-wp-tests.sh [hunk 14]
```diff
@@ -88,7 +88,7 @@ install_wp() {
 		else
 			local ARCHIVE_NAME="wordpress-$WP_VERSION"
 		fi
-		download https://wordpress.org/${ARCHIVE_NAME}.tar.gz  $TMPDIR/wordpress.tar.gz
+		download https://wordpress.org/${ARCHIVE_NAME}.tar.gz $TMPDIR/wordpress.tar.gz
```
Comment (aaemnnosttv) [inline tests/bin/install-wp-tests.sh:91]: Why was this line changed? This should be a mirror of the "main" script which lives in wp-cli's scaffolding.

**Review by aaemnnosttv — COMMENTED** [review]
Thanks @tofumatt ! Leaving a few comments here to provide some context for our discussion 😄

File: .github/workflows/php-tests.yml (hunk 6 above)
Comment (tofumatt) [reply .github/workflows/php-tests.yml:28]: Argh, sure, copypasta from another test getting 5.6 working 😅

File: .github/workflows/php-tests.yml (hunk 11 above)
Comment (tofumatt) [reply .github/workflows/php-tests.yml]: This is to prevent these from running at all for combos of the matrix we don't run (on Travis). But I'll tidy it up...

File: .github/workflows/php-tests.yml (hunk 13 above)
Comment (tofumatt) [reply .github/workflows/php-tests.yml]: See above: to skip it for combos of matrix values we want to skip.

File: tests/bin/install-wp-tests.sh (hunk 14 above)
Comment (tofumatt) [reply tests/bin/install-wp-tests.sh:91]: Oh, I had an error from something unrelated, noticed the double space, and might've just been tidying it. But if it's a mirror I'll restore 😉

File: .github/workflows/php-tests.yml (hunk 7 above)
Comment (tofumatt) [reply .github/workflows/php-tests.yml:31]: Urgh, right, sorry. 😓

File: .github/workflows/php-tests.yml (hunk 13 above)
Comment (aaemnnosttv) [reply .github/workflows/php-tests.yml]: Better to simplify the matrix and add includes for additional values that you don't want to generate permutations for.

**Comment by felixarntz** [comment]
@tofumatt I've removed `.travis.yml` and all Travis-CI references as part of this PR, since it's the last piece to migrate. Can you split out the PHP tests into separate jobs and then put each one into its own `.yml` file per scenario as discussed? This way it's aligned with the other tests.

Super excited to have everything running on GH Actions!

File: .github/workflows/php-tests-wp-4-7-php-5-6.yml [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+        env:
+          MYSQL_ALLOW_EMPTY_PASSWORD: yes
+        ports:
+          - 3306:3306
+        options: --health-cmd="mysqladmin ping" --health-interval=5s --health-timeout=2s --health-retries=3
+    env:
+      DB_HOST: 127.0.0.1
+      DB_PORT: 3306
+      MYSQL_USER: user
+      MYSQL_PASSWORD: password
```
Comment (eugene-manuilov) [inline .github/workflows/php-tests-wp-4-7-php-5-6.yml:30]:
```suggestion
      MYSQL_USER: root
      MYSQL_PASSWORD: root
```

File: .github/workflows/php-tests-wp-4-7-php-5-6.yml [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+    name: WP 4.7, PHP 5.6
+    runs-on: ubuntu-latest
+    services:
+      mysql:
+        image: mysql:5.7
+        env:
+          MYSQL_ALLOW_EMPTY_PASSWORD: yes
+        ports:
+          - 3306:3306
+        options: --health-cmd="mysqladmin ping" --health-interval=5s --health-timeout=2s --health-retries=3
```
Comment (eugene-manuilov) [inline .github/workflows/php-tests-wp-4-7-php-5-6.yml:25]: @tofumatt MySQL service isn't needed because there is a MySQL server on the virtual machine itself. Just run sudo systemctl start mysql.service to start it and you can use it with root user and root password.

File: .github/workflows/php-tests-wp-4-7-php-5-6.yml [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+          while ! mysqladmin ping -h"$DB_HOST" -P"$DB_PORT" --silent; do
+            sleep 1
+          done
+      - name: Grant database access
+        env:
+          MYSQL_TCP_PORT: ${{ env.DB_PORT }}
+        run: |
+          mysql -u root -h ${DB_HOST} -e "CREATE USER ${MYSQL_USER} IDENTIFIED BY '${MYSQL_PASSWORD}';"
+          mysql -u root -h ${DB_HOST} -e "GRANT ALL PRIVILEGES ON *.* TO ${MYSQL_USER};"
+          mysql -u root -h ${DB_HOST} -e "FLUSH PRIVILEGES;"
```
Comment (eugene-manuilov) [inline .github/workflows/php-tests-wp-4-7-php-5-6.yml:47]: This is not needed.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @tofumatt! This is a good start, but we need to adjust it. I have added a few comments to one of the new workflows, the same applies to others. Could you please take a look at it?

File: .github/workflows/php-tests.yml (hunk 7 above)
Comment (eugene-manuilov) [reply .github/workflows/php-tests.yml]: @aaemnnosttv the version of MySQL doesn't matter if it is 5.6 or higher. Why do you want to control its version?

File: .github/workflows/php-tests-wp-4-7-php-5-6.yml [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+          restore-keys: |
+            ${{ runner.os }}-composer-
+      - name: Composer Install
+        run: docker run --rm -v "$PWD:/app" -v "${{ steps.composer-cache.outputs.dir }}:/tmp/cache" composer install --no-interaction --no-progress # Run inside docker so composer install doesn't error...
+      - name: Set up PHP test data
+        run: tests/bin/install-wp-tests.sh ${MYSQL_DATABASE} ${MYSQL_USER} ${MYSQL_PASSWORD} ${DB_HOST}:${DB_PORT} ${WP_VERSION}
+      - name: Run Unit Tests
+        run: composer test
+        env:
+          WP_DB_HOST: ${{ env.DB_HOST }}:${{ env.DB_PORT }}
```
Comment (aaemnnosttv) [inline .github/workflows/php-tests-wp-4-7-php-5-6.yml:69]: This variable isn't necessary unless you wanted to substitute it above in the place of `${DB_HOST}:${DB_PORT}` but I think it's better to keep that part as-is for now.

**Review by aaemnnosttv — COMMENTED** [review]
@tofumatt @eugene-manuilov – I left a few comments/replies to add onto the initial review here.

File: .github/workflows/php-tests-wp-4-7-php-5-6.yml (hunk 16 above)
Comment (eugene-manuilov) [reply .github/workflows/php-tests-wp-4-7-php-5-6.yml:25]: @aaemnnosttv I don't have such doubts and think that it should work without any problem. Anyway, we can try it with the available version of MySQL server on the host and if it doesn't work for WordPress 4.7, then switch back to using a service, ok?

Comment (aaemnnosttv) [reply .github/workflows/php-tests-wp-4-7-php-5-6.yml:25]: @eugene-manuilov My concern there is that while it may work today, unexpected changes to the MySQL version on the runner could break things in the future. The [requirements page](https://make.wordpress.org/hosting/handbook/handbook/server-environment/#database) shows it supports MySQL `8.0.x` – does that mean older versions might not work with MySQL 8.1?

There are already a few tickets around MySQL 8:
- https://core.trac.wordpress.org/ticket/51740
- https://core.trac.wordpress.org/ticket/49364

It seems like it would be a good idea to stick to the latest MySQL v5 for now.

Comment (eugene-manuilov) [reply .github/workflows/php-tests-wp-4-7-php-5-6.yml:25]: I wish we could try MySQL 8 for our tests anyway and use it if it doesn't cause problems.

File: composer.lock [hunk 19]
```diff
@@ -1578,6 +1578,7 @@
             "keywords": [
                 "tokenizer"
             ],
+            "abandoned": true,
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply composer.lock:1581]: I don't think so—I was getting composer errors but they were unrelated to the lockfile, so reverted 👍🏻

File: .github/workflows/php-tests-wp-4-7-php-5-6.yml [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+        run: |
+          echo "::set-output name=dir::$(composer config cache-files-dir)"
+      - uses: actions/cache@v2
+        with:
+          path: ${{ steps.composer-cache.outputs.dir }}
+          key: ${{ runner.os }}-composer-${{ hashFiles('**/composer.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-composer-
+      - name: Composer Install
+        run: docker run --rm -v "$PWD:/app" -v "${{ steps.composer-cache.outputs.dir }}:/tmp/cache" composer install --no-interaction --no-progress # Run inside docker so composer install doesn't error...
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply .github/workflows/php-tests-wp-4-7-php-5-6.yml:63]: Yeah I think in this case we just want the install to work, and the speed improvements here are enough to use the docker command instead. Agreed it's less clean, but the comment at least clarifies the purpose 🙂

File: .github/workflows/php-tests-wp-4-7-php-5-6.yml (hunk 16 above)
Comment (tofumatt) [reply .github/workflows/php-tests-wp-4-7-php-5-6.yml:25]: The existing ones, I thought, assumed v5, so let's stick with that for now but add a v8 test (or move to it) in another issue? I'm just wary of introducing changes to the tests here during the migration.

File: .github/workflows/php-tests-wp-4-7-php-5-6.yml [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+      - name: Grant database access
+        env:
+          MYSQL_TCP_PORT: ${{ env.DB_PORT }}
+        run: |
+          mysql -u root -h ${DB_HOST} -e "CREATE USER ${MYSQL_USER} IDENTIFIED BY '${MYSQL_PASSWORD}';"
+          mysql -u root -h ${DB_HOST} -e "GRANT ALL PRIVILEGES ON *.* TO ${MYSQL_USER};"
+          mysql -u root -h ${DB_HOST} -e "FLUSH PRIVILEGES;"
+      - uses: shivammathur/setup-php@v2
+        with:
+          extensions: composer, mysqli
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply .github/workflows/php-tests-wp-4-7-php-5-6.yml:50]: Oh, shoot, you're right, I thought I needed it for set up but the docker path worked. Either this isn't needed, or I could move this to `tools` and save the `docker run` command…

File: .github/workflows/php-tests-wp-4-7-php-5-6.yml (hunk 17 above)
Comment (tofumatt) [reply .github/workflows/php-tests-wp-4-7-php-5-6.yml]: Trying to use the root user always results in an error for me, so I think we need this:

Screenshot 2020-12-21 at 19 20 36

**Review by felixarntz — APPROVED** [review]
Awesome!
